File: 01-Alex-Nicolaidis/wk01/5-fri/mta/js/mta.js
//MTA

//stations array for each line
const lineN = [ 'Times Square', '34th', '28th', '23rd', 'Union Square', '8th' ];
const lineL = [ '8th', '6th', 'Union Square', '3rd', '1st' ];
const line6 = [ 'Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place'];

//create a factory to generate the train lines and station

const trainFactory = function (l,s){
    return {
        line: l,
        stations:s
    }
};

const mtaTrains = [trainFactory("N",lineN), trainFactory("L",lineL), trainFactory("6",line6)];

const getStations = function (singleLine){   
    let train = mtaTrains.filter (function(a){
        return a.line === singleLine;
    });
    return train[0].stations;  //return an array of stations for a given line
};

// function to determine the stops if it is going forward
const forwardLoop = function (stationsList, start, end){
    let stopsArray = [];
    for(i = stationsList.indexOf(start) + 1; i < stationsList.indexOf(end) + 1; i++){
        stopsArray.push((stationsList[i]));
    }
    return stopsArray;
};

// function to determine the stops if it is going backwards
const backwardsLoop = function (trainLine, start, end){
    let stopsArray = [];
    for (i = trainLine.indexOf(start) - 1; i > trainLine.indexOf(end) - 1; i--){
        stopsArray.push(trainLine[i]);
    }
    return stopsArray;
};

const tripLeg = function ( trainLine, startStation, endStation ){
    let stops;
    let stations = getStations(trainLine);
    if(stations.indexOf(startStation) < stations.indexOf(endStation)){
        stops = forwardLoop(stations,startStation, endStation);
    }else{
        stops = backwardsLoop(stations,startStation, endStation);
    }
    return stops;
};

const planTrip = function (startLine, startStation, endLine, endStation){
    let totalStops = 0;
    if(startLine === endLine){
        let stops = tripLeg(startLine,startStation,endStation);
        totalStops += stops.length;
        console.log(`You must travel through the following stops on the ${startLine} line: ${stops.join(', ')}`);
        console.log(`${totalStops} stops in total`)
    }else{
        let stops1 = tripLeg(startLine,startStation,"Union Square");
        let stops2 = tripLeg(endLine,"Union Square",endStation);
        totalStops += stops1.length + stops2.length;
        console.log(`You must travel through the following stops on the ${startLine} line: ${stops1.join(', ')}`);
        console.log("Change at Union Sqaure");
        console.log(`Your journey continues through the following stops: ${stops2.join(', ')} on the ${endLine}.`);
        console.log(`${totalStops} stops in total`)
    };
};


planTrip('N', 'Times Square', '6', '33rd');

File: 01-Alex-Nicolaidis/wk04/2-tue/mta.rb
require "pry"
#MTA attempt 2
#sort the train lines in a hash
@lines = {
    :N => ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
    :L => ["8th", "6th", "Union Square", "3rd", "1st"],
    :"6" => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "8th"]
}

#Work out how to travel on 1 line
starting_line = "N"
start_station = "Times Square"
ending_line = "6"
end_station = "33rd"

def trip_leg line, start, stop
    line_stations = @lines[line.to_sym] #convert input string to a symbol
    #find the index of the start and stop position in the train list arary
    start_index = line_stations.find_index(start)
    stop_index = line_stations.find_index(stop)
    #check if we are going forward or backwards
    if start_index < stop_index
        line_stations[start_index+1..stop_index]
    else
        reverse_stations = line_stations.reverse
        reverse_stations[reverse_stations.find_index(start)+1 ..reverse_stations.find_index(stop)]
    end
end

def plan_trip start_line, start_station, end_line, end_station
    if start_line == end_line
        trip_array = trip_leg start_line, start_station, end_station
        puts "Your must travel through the following stops on the #{ start_line }: #{ trip_array.join(", ")}."
        puts "#{trip_array.size} stops in total"
    else
        change_station = "Union Square"
        leg_1 = trip_array = trip_leg start_line, start_station, change_station
        leg_2 = trip_array = trip_leg end_line, change_station, end_station
        total_stops = (leg_1 + leg_2).size
        puts "Your must travel through the following stops on the #{ start_line }: #{ leg_1.join(", ")}."
        puts "Change at Union Square"
        puts "Your journey continues through the following stations on the #{ end_line }: #{ leg_2.join(", ")}."
        puts "#{ total_stops } stops in total"
    end
end

plan_trip starting_line, start_station, ending_line, end_station





